src/Compilers/Core/Portable/FileSystem/PathUtilities.cs

```diff
@@ -773,6 +773,31 @@ public static bool IsValidFilePath([NotNullWhen(true)] string? fullPath)
         public static string NormalizeWithForwardSlash(string p)
             => DirectorySeparatorChar == '/' ? p : p.Replace(DirectorySeparatorChar, '/');
 
+        /// <summary>
+        /// Takes a normalized full path from <see cref="NormalizeWithForwardSlash(string)"/> and expands any '..' or '.'
+        /// into their equivalent representation.
+        /// </summary>
+        /// <returns>An equivalent path that does not contain any '..' or '.' path parts.</returns>
+        public static string ExpandNormalizedFullPathWithRelativeParts(string p)
+        {
+            Debug.Assert(p.StartsWith("/"));
+
+            var parts = GetPathParts(p);
+            Stack<string> resolvedParts = new Stack<string>();
+            foreach (var part in parts)
+            {
+                if (!part.Equals(ParentRelativeDirectory))
+                {
+                    resolvedParts.Push(part);
+                }
+                else if (resolvedParts.Count > 0)
```

Comment:
Is it fine to ignore the `ParentRelativeDirectory` path component if it is not preceded by some "normal" path component?

Reply:
Yeah, at least on windows and linux `\..\..\bin` == `\bin`. So if we see a relative directory component with no preceding parent, we can essentially just ignore it.